src/Core/MailTemplate/MailLayoutFolderCatalog.php

```diff
@@ -0,0 +1,246 @@
+            ]
+        );
+        $this->checkThemeFolder($mailThemeFolder);
+
+        $mailThemeLayouts = new MailLayoutCollection();
+        $this->listCoreLayouts($mailThemeLayouts, $mailThemeFolder);
+        $this->listModulesLayouts($mailThemeLayouts, $mailThemeFolder);
+
+        //This hook allows to add/remove layouts to a mail theme
+        $this->hookDispatcher->dispatchWithParameters(
+            MailLayoutCatalogInterface::LIST_MAIL_THEME_LAYOUTS_HOOK,
+            [
+                'mailTheme' => $mailTheme,
+                'mailThemeLayouts' => $mailThemeLayouts,
+            ]
+        );
+
+        return $mailThemeLayouts;
+    }
+
+    /**
+     * @param MailLayoutCollectionInterface $collection
+     * @param string $mailThemeFolder
+     */
+    private function listCoreLayouts(MailLayoutCollectionInterface $collection, $mailThemeFolder)
+    {
+        $coreLayoutsFolder = implode(DIRECTORY_SEPARATOR, [
+            $mailThemeFolder,
+            MailTemplateInterface::CORE_CATEGORY,
+        ]);
+        if (!is_dir($coreLayoutsFolder)) {
+            return;
+        }
+
+        $this->addLayoutsFromFolder($collection, $coreLayoutsFolder);
+    }
+
+    /**
+     * @param MailLayoutCollectionInterface $collection
+     * @param string $mailThemeFolder
+     */
+    private function listModulesLayouts(MailLayoutCollectionInterface $collection, $mailThemeFolder)
+    {
+        $moduleLayoutsFolder = implode(DIRECTORY_SEPARATOR, [
+            $mailThemeFolder,
+            MailTemplateInterface::MODULES_CATEGORY,
+        ]);
+        if (!is_dir($moduleLayoutsFolder)) {
+            return;
+        }
+
+        $moduleFinder = new Finder();
+        $moduleFinder->directories()->in($moduleLayoutsFolder)->depth(0);
+
+        /* @var SplFileInfo $fileInfo */
+        foreach ($moduleFinder as $moduleFileInfo) {
+            $moduleName = $moduleFileInfo->getFilename();
+            $moduleFolder = implode(DIRECTORY_SEPARATOR, [$moduleLayoutsFolder, $moduleName]);
+            $this->addLayoutsFromFolder($collection, $moduleFolder, $moduleName);
+        }
+    }
+
+    /**
+     * @param MailLayoutCollectionInterface $collection
+     * @param string $folder
+     * @param string $moduleName
+     */
+    private function addLayoutsFromFolder(
+        MailLayoutCollectionInterface $collection,
+        $folder,
+        $moduleName = ''
+    ) {
+        $layoutFiles = [];
+        $finder = new Finder();
+        $finder->files()->in($folder)->sortByName();
+        /** @var SplFileInfo $fileInfo */
+        foreach ($finder as $fileInfo) {
+            //Get filename without any extension (ex: account.html.twig -> account)
+            $layoutName = preg_replace('/\..+/', '', $fileInfo->getBasename());
+            if (!isset($layoutFiles[$layoutName])) {
+                $layoutFiles[$layoutName] = [
+                    MailTemplateInterface::HTML_TYPE => '',
```

Comment:
should it be initialized to null?

Reply:
I thought about it, but as @mickaelandrieu suggested I prefer to keep a single type, so I use empty string rather than null. (I don't always agree with this principal, but in this case it seemed appropriate)